spec/requests/vms_spec.rb

```diff
@@ -2139,6 +2139,48 @@ def query_match_regexp(*tables)
     end
   end
 
+  describe "POST /api/vms/:id" do
+    let(:ems) { FactoryBot.create(:ems_openstack) }
+    let(:tenant) { FactoryBot.create(:cloud_tenant_openstack, :ext_management_system => ems) }
+    let(:floating_ip) { FactoryBot.create(:floating_ip_openstack, :ext_management_system => ems.network_manager, :cloud_tenant => tenant) }
+
+    it 'associates a floating ip to the vm' do
+      vm = FactoryBot.create(:vm_openstack, :ems_id => ems.id, :cloud_tenant => tenant)
+      api_basic_authorize action_identifier(:vms, :associate)
+
+      post(api_vms_url, :params => {:action => 'associate', :resource => {:id => vm.id, :floating_ip => floating_ip}})
+
+      expect(response).to have_http_status(:ok)
+      task_id = response.parsed_body["task_id"]
+      expected = {
+        "results" => [{
+          "message"   => "Associating resource to Vm id: #{vm.id} name: '#{vm.name}",
+          "success"   => true,
+          "task_id"   => "#{task_id}",
+        }]
+      }
+      expect_single_action_result(expected)
+    end
+
+    it 'disassociates a floating ip from the vm' do
+      vm = FactoryBot.create(:vm_openstack, :ems_id => ems.id, :cloud_tenant => tenant, :floating_ips => [floating_ip])
+      api_basic_authorize action_identifier(:vms, :disassociate)
+
+      post(api_vms_url, :params => {:action => 'disassociate', :resource => {:id => vm.id, :floating_ip => floating_ip}})
```

Comment:
we could try adding a test to add a floating ip to a vm with already one floating ip since we have a `has_one` and `has_many`  https://github.com/ManageIQ/manageiq/blob/4abf8c6ddd9f1985c1264ca407837423695b1dfd/app/models/manageiq/providers/cloud_manager/vm.rb#L12-L13)

Reply:
Ok, let me know when your changes are pushed or you have other ideas.  👍